### PR TITLE
karpenter 0.32

### PR DESCRIPTION
**Pull request by thepoppingone**
- resolve ip6 conflict
- Refactor karpenter
- Remove unused vars
- Update flag for count using autoscaling mode
- Add more details for allow CRDs installation first
- Update vars
- Resolve readme conflict
- Patch again to stop duplicate creation of logging policy
- Re apply karpenter module after CRD installed
- Try just use values
- set attribute accessor
- Fix service account missing role arn
- set chart version variable for karpenter
- Use cluster sec group instead of node as its using fargate
- Use 8 cpu
- Try empty string map instead of null object
- Set default tags
- Attach primary cluster sec group as well to nodes on karpenter
- Add outputs
- Use cluster primary sec group
- convert to kubectl
- Remove install crd first var flag
- Test if requirements is the error
- Test no taints and labels
- Fix missing taints and labels key
- Remove taints to test again
- Test provisioneres removal
- Test indent
- Add indent function
- Test removal of quotes
- Add indent
- Add depends on
- Make node templates required var
- remove default observability ns creation as default fargate profile would handle it
- Add kubectl provider
- Add helm provider version
- Add kubernetes provider
- Remove extra provider
- remove provider config
- add providers back
- Add block device mapping customiziation
- Test device block mapping
- add provisioner_name to template
- Add enablePodEni - defaults to true
- Use node label instead of taint
- Set correctly taints
- Fix logic on checking length
- Add secgroupol
- add depends on for helm installation to create ns first
- first draft
- first draft

### CODE REVIEW

**Comment by thepoppingone** [comment]
If i dont add the provider passing the tflint issues appear

File: karpenter.tf [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+  # Karpenter Nodetemplate Config
+  karpenter_nodeclasses = coalescelist(var.karpenter_nodeclasses, [
+    {
+      nodeclass_name = "default"
+      karpenter_subnet_selector_maps = [{
+        tags = {
+          "Name" = "aft-app-ap-southeast*"
+        },
+        }
```
Comment (niroz89) [inline karpenter.tf:14]: We shouldn't use this, better if we can use karpenter discovery tag.

`karpenter.sh/discovery: "${CLUSTER_NAME}"`

But this means, we need to allow adding subnet tag with a custom value using `aws_ec2_tag` and allow SCP. OR

we can use existing tags. ex:

`"kubernetes.io/role/internal-elb" = "1"`

**Comment by niroz89** [comment]
> If i dont add the provider passing the tflint issues appear

@thepoppingone added tflint ignore.

File: karpenter.tf (hunk 1 above)
Comment (uchinda-sph) [reply karpenter.tf]: introduce `karpenter_subnet_selector_tags` variable to pass both key and value of the tags 
```
karpenter_subnet_selector_tags = {
  "Name" = "App*"
}
```

**Review by uchinda-sph — APPROVED** [review]

File: variables.tf [hunk 2]
```diff
@@ -518,3 +513,11 @@ variable "karpenter_chart_version" {
   type        = string
   default     = "v0.32.1"
 }
+
+variable "karpenter_subnet_selector_tags" {
+  description = "Subnet selector tags for Karpenter nodes"
+  type        = map(string)
+  default = {
+    "Name" = "aft-app-ap-southeast*"
+  }
```
Comment (niroz89) [inline variables.tf:522]: Let's remove the `default` and make it mandatory.

Comment (uchinda-sph) [reply variables.tf]: Done

Comment (uchinda-sph) [reply variables.tf]: @niroz89 if we set this as mandatory then tf will fail when assign value to `var.karpenter_nodeclasses`
https://github.com/SPHTech-Platform/platform-engg-tools/blob/1c325f30697c8ce1abe058e192a1bb760e30486e/envs/prd/github_runner_enterprise/eks_pure_karpenter_cluster/main.tf#L111

**Review by niroz89 — APPROVED** [review]

**Pull request merged**